vassal-app/src/main/java/VASSAL/counters/Mat.java

```diff
@@ -0,0 +1,308 @@
+    return contents.contains(p);
+  }
+
+  /**
+   * Adds a piece of cargo to this mat
+   * @param p game piece to add
+   */
+  public void addCargo(GamePiece p) {
+    if (!(p instanceof Decorator) || hasCargo(p)) {
+      return;
+    }
+
+    contents.add(p);
+
+    final GamePiece cargo = Decorator.getDecorator(Decorator.getOutermost(p), MatCargo.class);
+    if (cargo != null) {
+      ((MatCargo)cargo).setMat(Decorator.getOutermost(this));
+    }
+  }
+
+
+  /**
+   * Adds a piece of cargo and returns a command to duplicate the operation on another client
+   * @param p piece of cargo to add
+   * @return Command that adds the cargo to this mat (and removes it from any other mat it was on)
+   */
+  public Command makeAddCargoCommand(GamePiece p) {
+    final ChangeTracker ct  = new ChangeTracker(this);
+    final ChangeTracker ct2 = new ChangeTracker(p);
+    ChangeTracker ct3 = null;
+
+    if ((p instanceof Decorator) && !hasCargo(p)) {
+      final GamePiece cargo = Decorator.getDecorator(Decorator.getOutermost(p), MatCargo.class);
+      if (cargo != null) {
+        final GamePiece mt = ((MatCargo)cargo).getMat();
+        if ((mt != null) && (mt != Decorator.getOutermost(this))) {
+          ct3 = new ChangeTracker(mt);
+        }
+      }
+    }
+
+    addCargo(p);
+
+    Command c = ct.getChangeCommand().append(ct2.getChangeCommand());
+    if (ct3 != null) {
+      c = c.append(ct3.getChangeCommand());
+    }
+
+    return c;
+  }
+
+  /**
+   * Removes a MatCargo piece from our list of cargo
+   * @param p Cargo to remove
+   */
+  public void removeCargo(GamePiece p) {
+    if ((p instanceof Decorator) && hasCargo(p)) {
+      contents.remove(p);
+      final GamePiece mp = Decorator.getDecorator(p, MatCargo.class);
+      if (mp != null) {
+        ((MatCargo)mp).clearMat();
+      }
+    }
+  }
+
+  /**
+   * Removes a MatCargo piece from our list of cargo, and returns a Command to duplicate the changes on another client
+   * @param p GamePiece with a MatCargo trait, to be removed
+   * @return Command to remove the piece
+   */
+  public Command makeRemoveCargoCommand(GamePiece p) {
+    final ChangeTracker ct  = new ChangeTracker(this);
+    final ChangeTracker ct2 = new ChangeTracker(p);
+
+    removeCargo(p);
+
+    return ct.getChangeCommand().append(ct2.getChangeCommand());
+  }
+
+
+  @Override
+  public Rectangle boundingBox() {
+    return piece.boundingBox();
+  }
+
+  @Override
+  public void draw(Graphics g, int x, int y, Component obs, double zoom) {
+    piece.draw(g, x, y, obs, zoom);
+  }
+
+  @Override
+  public String getName() {
+    return piece.getName();
+  }
+
+  @Override
+  public Shape getShape() {
+    return piece.getShape();
+  }
+
+  @Override
+  public PieceEditor getEditor() {
+    return new Ed(this);
+  }
+
+  @Override
+  public String getDescription() {
+    return buildDescription("Editor.Mat.trait_description", desc);
+  }
+
+  @Override
+  public Object getProperty(Object key) {
```

Comment:
Will need a property to expose number of cargo pieces this mat holds. 

Probably also need the equivalent of Deck  'Count Expressions' and the properties to expose them, keyed off the mat name.